### PR TITLE
Convert plugin javascript code to typescript

### PR DESCRIPTION
**Pull request by TomBarzilay1**
This PR converts plugin js code to typescript.

1. `cd project/plugins; npm run build` to generate type declarations and minified js bundle
2. Added interfaces for all models accessible via API
3. Align data returned by native platforms to js. Aligned with Flutter SDK

TO COME: convert example app to ts, verify all interfaces usable as expected.

### CODE REVIEW

File: plugins/package.json [hunk 1]
```diff
@@ -25,6 +26,15 @@
     ]
   },
   "devDependencies": {
-    "cordova-ios": "^6.2.0"
+    "@babel/preset-typescript": "^7.18.6",
+    "babel": "^6.23.0",
+    "babel-core": "^6.26.3",
+    "babel-loader": "^7.1.5",
+    "cordova-ios": "^6.2.0",
+    "source-map-loader": "^4.0.0",
```
Comment (vvoicehovics) [inline plugins/package.json:34]: are all of the dependencies here used ? For example, `source-map-loader` or `babel-loader` are not in webpack config. Or they are required by other dependencies? Could you describe what are they for and how applied

File: plugins/src/core/ddl.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+interface DeepLinkContent {
+  title?: string;
+  description?: string;
+}
+
+export interface DeepLink {
+  content: DeepLinkContent;
+  data: any;
```
Comment (vvoicehovics) [inline plugins/src/core/ddl.ts:8]: data is a json object, mb can be a bit more specific with type

File: plugins/package.json (hunk 1 above)
Comment (TomBarzilay1) [reply plugins/package.json:34]: these dependencies were auto generated in one of the iterations of trying to make the bundling work, I will delete them

File: plugins/src/core/ddl.d.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+interface DeepLinkContent {
+    title?: string;
+    description?: string;
+}
+export interface DeepLink {
+    content: DeepLinkContent;
+    data: any;
+    url: string;
+}
+export {};
```
Comment (vvoicehovics) [inline plugins/src/core/ddl.d.ts:10]: ?

File: plugins/src/core/index.ts [hunk 4]
```diff
@@ -0,0 +1,369 @@
+import * as cordova from "cordova";
+import {
+  PushNotificationHandler,
+  InAppDeepLinkHandler,
+  InAppInboxUpdatedHandler,
+  DeepLinkHandler,
+} from "./handlers";
+
+export interface OptimoveConfig{
+  pushReceivedHandler?: PushNotificationHandler | null;
```
Comment (vvoicehovics) [inline plugins/src/core/index.ts:10]: are these not optional? They are always null or a function

File: plugins/src/core/index.ts [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+
+interface InAppInboxItem {
+  id: number;
+  title: string;
+  subtitle: string;
+  availableFrom: string | "";
+  availableTo: string | "";
+  dismissedAt: string | "";
+  isRead: boolean;
+  sentAt?: string;
```
Comment (vvoicehovics) [inline plugins/src/core/index.ts:121]: it's not enough to copy from the old SDK, ideally need to verify that types in Optimove SDKs match . for example sentAt i think is not optional?

File: plugins/src/core/index.ts [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+      nativeMessageHandler,
+      reject,
+      "OptimoveSDKPlugin",
+      "checkIfPendingPushExists",
+      []
+    );
+  });
+}
+
+function nativeMessageHandler(message) {
```
Comment (vvoicehovics) [inline plugins/src/core/index.ts:91]: can add type for the arg

File: plugins/src/core/index.ts [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+  if (
+    handlerName === "inAppInboxUpdatedHandler" &&
+    typeof currentConfig[handlerName] === "function"
+  ) {
+    currentConfig[handlerName]?.();
+
+    return;
+  }
+
+  if (typeof currentConfig[handlerName] === "function") {
```
Comment (vvoicehovics) [inline plugins/src/core/index.ts:106]: can be replaced with `currentConfig[handlerName] !== null`. And same for all typeof

File: plugins/src/core/index.ts [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+  [key: string]: any;
+}
+interface InAppInboxSummary {
+  totalCount: number;
+  unreadCount: number;
+}
+
+const Optimove = {
+  /**
+   * Sets the User ID of the current user and starts the {@code Visitor} to {@code Customer} conversion flow.
```
Comment (vvoicehovics) [inline plugins/src/core/index.ts:135]: are these descriptions taken from somewhere? Do consumers of the SDK really need to know about `starts the {@code Visitor} to {@code Customer} conversion flow.`?

File: plugins/src/core/ddl.ts (hunk 2 above)
Comment (TomBarzilay1) [reply plugins/src/core/ddl.ts:8]: image
 I took this object as a reference, but of course when I look at the actual object that we send from the native function the objects are different, I'll fix this now.

File: plugins/src/core/index.ts (hunk 8 above)
Comment (TomBarzilay1) [reply plugins/src/core/index.ts:135]: I used Android SDK and Kumulos Cordova SDK as a reference

File: plugins/src/core/push.ts [hunk 9]
```diff
@@ -0,0 +1,7 @@
+export interface PushNotification {
+  id: number;
+  title?: string | null;
+  message?: string | null;
+  data?: { [key: string]: any } | null;
```
Comment (vvoicehovics) [inline plugins/src/core/push.ts:5]: same as above. consistency with cordova ios side/Flutter, verify it's optional VS null

File: plugins/src/core/ddl.d.ts (hunk 3 above)
Comment (TomBarzilay1) [reply plugins/src/core/ddl.d.ts:10]: this file  was created automatically after the bundling, I'm not sure why the export {}; is needed, seems weird, I'll delete that.

File: plugins/src/core/index.ts (hunk 8 above)
Comment (vvoicehovics) [reply plugins/src/core/index.ts:135]: y, but java annotations dont belong to typescript

File: plugins/src/core/index.ts (hunk 4 above)
Comment (TomBarzilay1) [reply plugins/src/core/index.ts]: changed the handlers to non - optional

File: plugins/src/core/index.ts (hunk 5 above)
Comment (TomBarzilay1) [reply plugins/src/core/index.ts]: I aligned the interface with flutter sdk

File: plugins/src/core/handlers.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import { PushNotification } from "./push";
+import { DeepLink } from "./ddl";
+export interface PushNotificationHandler {
+  (notification: PushNotification): void;
+}
+
+export interface InAppDeepLinkHandler {
+  (data: { [key: string]: any }): void;
```
Comment (vvoicehovics) [inline plugins/src/core/handlers.ts:8]: InAppButtonPress is the argument, which is 
```
public struct InAppButtonPress {
    public let deepLinkData : [AnyHashable:Any]
    public let messageId : Int64
    public let messageData : NSDictionary?

    init(deepLinkData:[AnyHashable:Any], messageId:Int64, messageData:NSDictionary?) {
        self.deepLinkData = deepLinkData
        self.messageId = messageId
        self.messageData = messageData
    }
}
```

in my test i've received this `{"messageData":null,"deepLinkData":{"hellow":"win"},"messageId":119}`

File: Example_App/www/js/index.js [hunk 11]
```diff
@@ -249,12 +247,4 @@ function inAppDeleteMessageFromInbox() {
     error
   );
 }
-document
```
Comment (vvoicehovics) [inline Example_App/www/js/index.js:252]: removed listener?

File: Example_App/www/js/index.js [hunk 12]
```diff
@@ -249,12 +247,4 @@ function inAppDeleteMessageFromInbox() {
...
-  .getElementById("get-visitor-id-button")
-  .addEventListener("click", getVisitorId);
```
Comment (vvoicehovics) [inline Example_App/www/js/index.js:254]: removed listener?

File: plugins/src/core/ddl.ts (hunk 2 above)
Comment (vvoicehovics) [reply plugins/src/core/ddl.ts]: resolution can be an enum

File: plugins/src/core/index.ts (hunk 5 above)
Comment (vvoicehovics) [reply plugins/src/core/index.ts]: aaaa, this is not `imagePath`, it's `imageUrl`. and availableTo/From are "", not null. Can double check this again?

File: plugins/src/core/ddl.ts [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+}
+ 
+export interface DeepLink {
+  link: string;
+  resolution: DeepLinkResolution;
+  data: DeepLinkData;
+}
+
+export enum DeepLinkResolution {
+  LOOKUP_FAILED,
```
Comment (vvoicehovics) [inline plugins/src/core/ddl.ts:18]: enum values here are strings, not integers

File: plugins/src/core/inApp.ts [hunk 14]
```diff
@@ -0,0 +1,23 @@
+export interface InAppInboxItem {
+  id: number;
+  title: string;
+  subtitle: string;
+  availableFrom: string | null; // Date?
+  availableTo: string | null;
+  dismissedAt: string | null;
```
Comment (vvoicehovics) [inline plugins/src/core/inApp.ts:7]: availableFrom, availableTo and dismissedAt are being passed as "" from native. nulls are what Flutter does. Can we align native with js? you can add/perhaps even test in android. ill add in swift

File: plugins/src/core/index.ts (hunk 8 above)
Comment (vvoicehovics) [reply plugins/src/core/index.ts]: same for @link

File: plugins/src/core/push.ts [hunk 15]
```diff
@@ -0,0 +1,7 @@
+export interface PushNotification {
+  id: number;
+  title: string | null;
+  message: string | null;
+  data: { [key: string]: any } | null;
```
Comment (vvoicehovics) [inline plugins/src/core/push.ts:5]: can we use `JSON | null` like in other places

Comment (vvoicehovics) [reply plugins/src/core/push.ts:5]: replaced JSON with `{ [key: string]: any }`

File: plugins/src/core/ddl.ts [hunk 16]
```diff
@@ -0,0 +1,23 @@
+export interface DeepLinkContent {
+  title?: string;
+  description?: string;
+}
+export interface DeepLinkData {
+  data: { [key: string]: any };
```
Comment (cgwyllie) [inline plugins/src/core/ddl.ts:6]: Can possibly use `Record<string,any>` for these types of declarations?

File: plugins/tsconfig.json [hunk 17]
```diff
@@ -0,0 +1,19 @@
+{
+  "compilerOptions": {
+        "target": "es2015",
+        "allowSyntheticDefaultImports": true,
+        "declaration": true,
+        "outDir": "./src/core/declarations",
+        "emitDecoratorMetadata": true,
+        "experimentalDecorators": true,
```
Comment (cgwyllie) [inline plugins/tsconfig.json:8]: Is this 'standard'?

File: plugins/tsconfig.json [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+        "rootDir": "./src/core",
+        "module": "esnext",
+        "moduleResolution": "node"
+    },
+     "include":[
+        "src/core/*.ts"
```
Comment (cgwyllie) [inline plugins/tsconfig.json:14]: Can/should we glob files to allow future module additions more easily?

e.g. `src/**/*.ts`

File: plugins/package.json [hunk 19]
```diff
@@ -5,6 +5,7 @@
   "description": "A sample Apache Cordova application that responds to the deviceready event.",
   "main": "index.js",
```
Comment (cgwyllie) [inline plugins/package.json:6]: I think we can/should perhaps declare the "types" field to point at index.d.ts for the module?

Maybe it's automatically picked up based on some convention, but if in a non-standard location I think we have to declare it.

File: plugins/plugin.xml [hunk 20]
```diff
@@ -1,6 +1,7 @@
 <?xml version='1.0' encoding='utf-8'?>
 <plugin id="optimove-cordova-plugin" version="1.0.0" xmlns="http://apache.org/cordova/ns/plugins/1.0" xmlns:android="http://schemas.android.com/apk/res/android">
     <name>OptimoveSDKPlugin</name>
+    <js-module name="OptimoveCore" src="www/OptimoveCore.js"/>
     <js-module name="Optimove" src="www/Optimove.js">
```
Comment (cgwyllie) [inline plugins/plugin.xml:5]: Do we need to declare both modules / what's the reason for the split?

Comment (cgwyllie) [reply plugins/plugin.xml:5]: (ah sorry see this is how we have it set up in Kumulos plugin too)

File: plugins/src/core/ddl.ts (hunk 16 above)
Comment (vvoicehovics) [reply plugins/src/core/ddl.ts]: updated. what's the benefit, only slightly neater syntax?

File: plugins/plugin.xml (hunk 20 above)
Comment (vvoicehovics) [reply plugins/plugin.xml:5]: i think it's possible to output a module from webpack immediately https://webpack.js.org/configuration/output/#outputmodule . But it's an experimental feature and i figured easier to not mess up with it

File: plugins/package.json (hunk 19 above)
Comment (vvoicehovics) [reply plugins/package.json:6]: yeah,  "main" is quite certainly incorrect. Im not sure if my change is 100% right, in the old SDK we used to copy some of the types into root folder for some reason.

Maybe can verify types are found when typescriptify the app

File: plugins/src/core/ddl.ts (hunk 16 above)
Comment (cgwyllie) [reply plugins/src/core/ddl.ts]: Yeah, just a nice utility type and a bit easier to type out (I think).

https://www.typescriptlang.org/docs/handbook/utility-types.html

File: plugins/package.json (hunk 19 above)
Comment (cgwyllie) [reply plugins/package.json:6]: TS type declarations should also help (if they are found) in non-TS apps the are edited in a TS def aware editor like VSCode. They typing info should be found in the JS app & help w/ arg hints etc. if it's working properly I think.

Comment (cgwyllie) [reply plugins/package.json:6]: Pick this up as follow-up.

**Review by cgwyllie — APPROVED** [review]

**Pull request merged**